apps/frontend/src/context/DataContext.tsx

```diff
@@ -60,43 +60,59 @@ export const DataContextProvider = ({ clockWorker, children }: Props) => {
   const { heartRate } = useHeartRateMonitor();
 
   const hasValidData = trackedData.length > 0;
+
+  const accumulatedTimeRef = React.useRef(0);
+
   const dataTick = React.useCallback(
-    (
+    async (
       delta: number,
       data: {
         heartRate: number | null;
         power: number | null;
         cadence: number | null;
       }
     ) => {
-      const heartRateToInclude = data.heartRate
-        ? { heartRate: data.heartRate }
-        : {};
-      const powerToInclude = data.power ? { power: data.power } : {};
-      const cadenceToInclude = data.cadence ? { cadence: data.cadence } : {};
-      const dataPoint = {
-        timestamp: new Date(),
-        ...heartRateToInclude,
-        ...powerToInclude,
-        ...cadenceToInclude,
-      };
-
-      if (dataPoint.cadence || dataPoint.power || dataPoint.heartRate) {
-        db.addDatapoint(delta, data, isRunning);
-        sendData(dataPoint);
+      try {
+        if (accumulatedTimeRef.current > 0) {
+          await db.addElapsedTime(accumulatedTimeRef.current);
+          accumulatedTimeRef.current = 0;
+        }
+
+        const heartRateToInclude = data.heartRate
+          ? { heartRate: data.heartRate }
+          : {};
+        const powerToInclude = data.power ? { power: data.power } : {};
+        const cadenceToInclude = data.cadence ? { cadence: data.cadence } : {};
+        const dataPoint = {
+          timestamp: new Date(),
+          ...heartRateToInclude,
+          ...powerToInclude,
+          ...cadenceToInclude,
+        };
+
+        if (dataPoint.cadence || dataPoint.power || dataPoint.heartRate) {
+          await db.addDatapoint(delta, data, isRunning);
+          sendData(dataPoint);
+        }
+      } catch (error) {
+        console.error('Failed to process dataTick:', error);
       }
     },
-    [heartRate, power, cadence, isRunning]
+    [isRunning, sendData]
+  );
+
+  const clockTick = React.useCallback(
+    (delta: number) => {
+      accumulatedTimeRef.current += delta;
+      increaseActiveWorkoutElapsedTime(delta);
+    },
+    [increaseActiveWorkoutElapsedTime]
   );
-  const clockTick = (delta: number) => {
-    db.addElapsedTime(delta);
```

Comment:
The removal of this is the key part of this PR. We're increasing the ref instead, and calling `db.addElapsedTime` on the data tick